src/python/pants/engine/internals/dep_rules.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Iterable, Mapping, Sequence, Tuple, Union
+
+from typing_extensions import Literal
+
+from pants.engine.internals.target_adaptor import TargetAdaptor
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.unions import UnionMembership, union
+
+logger = logging.getLogger(__name__)
+
+
+SetDependencyRulesValueT = Tuple[str, ...]
+SetDependencyRulesKeyT = Union[str, Tuple[str, ...]]
+SetDependencyRulesT = Mapping[SetDependencyRulesKeyT, SetDependencyRulesValueT]
+
+
+class DependencyRulesError(Exception):
+    pass
+
+
+class DependencyRuleActionDeniedError(DependencyRulesError):
+    pass
+
+
+class DependencyRuleAction(Enum):
+    ALLOW = "allow"
+    DENY = "deny"
+    WARN = "warn"
+
+    def execute(self, *, description_of_origin: str) -> None:
+        if self is DependencyRuleAction.ALLOW:
+            return
+        msg = f"Dependency rule violation for {description_of_origin}"
+        if self is DependencyRuleAction.DENY:
+            raise DependencyRuleActionDeniedError(msg)
+        if self is DependencyRuleAction.WARN:
+            logger.warning(msg)
+
+
+SetDefaultDependencyRulesT = Union[Literal["allow"], Literal["deny"], Literal["warn"]]
+
+
+class DependencyRule:
+    pass
+
+
+DependencyRules = Tuple[DependencyRule, ...]
+
+
+class BuildFileDependencyRules(ABC):
+    default: DependencyRuleAction
+    all: DependencyRules
+    targets: Mapping[str, DependencyRules]
+
+    @classmethod
+    @abstractmethod
+    def create(
+        cls,
+        default: DependencyRuleAction = DependencyRuleAction.ALLOW,
+        all: Iterable[str | DependencyRule] = (),
+        targets: Mapping[str, Iterable[str | DependencyRule]] = {},
+    ) -> BuildFileDependencyRules:
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def check_dependency_rules(
+        *,
+        source_adaptor: TargetAdaptor,
+        source_path: str,
+        dependencies_rules: BuildFileDependencyRules | None,
+        target_adaptor: TargetAdaptor,
+        target_path: str,
+        dependents_rules: BuildFileDependencyRules | None,
+    ) -> DependencyRuleAction:
```

Comment:
Changed the signature slightly from passing the target `type_alias` to passing the full `TargetAdaptor` instead. The idea being to offer the backend implementation a little more information to go on than merely the target type and path. Not super convinced if this is a good move, though. (the motivation originates from an idea that it could be possible to have rules apply to targets based on their `tags`, and for stuff like that we would need this change..)

Reply:
Yeah, I think that makes sense. The `tags` thing is going to come up sooner or later :)